modules/gapi/perf/common/gapi_render_perf_tests.cpp

```diff
@@ -0,0 +1,1281 @@
+
+  class RenderTestLines : public TestPerfParams<tuple<cv::Size, cv::Point, cv::Point,
+                                                      cv::Scalar, int, int, int, MatType>> {};
+
+  class RenderTestMosaics : public TestPerfParams<tuple<cv::Size, cv::Rect, int, int,
+                                                        MatType>> {};
+
+  class RenderTestImages : public TestPerfParams<tuple<cv::Size, cv::Rect, cv::Scalar, double,
+                                                       MatType>> {};
+
+  class RenderTestPolylines : public TestPerfParams<tuple<cv::Size, std::vector<cv::Point>,
+                                                          cv::Scalar, int, int, int, MatType>> {};
+
+  class RenderTestPolyItems : public TestPerfParams<tuple<cv::Size, int, int, int>> {};
+
+
+namespace {
+void create_rand_mats(const cv::Size &size, MatType type, cv::Mat &ref_mat, cv::Mat &gapi_mat)
+{
+    ref_mat.create(size, type);
+    cv::randu(ref_mat, cv::Scalar::all(0), cv::Scalar::all(255));
+    ref_mat.copyTo(gapi_mat);
+};
+
+// FIXME: remove this code duplication with /gapi/test/common/gapi_render_tests.hpp
+cv::Scalar cvtBGRToYUVC(const cv::Scalar &bgr)
+{
+    double y = bgr[2] * 0.299000 + bgr[1] * 0.587000 + bgr[0] * 0.114000;
+    double u = bgr[2] * -0.168736 + bgr[1] * -0.331264 + bgr[0] * 0.500000 + 128;
+    double v = bgr[2] * 0.500000 + bgr[1] * -0.418688 + bgr[0] * -0.081312 + 128;
+    return {y, u, v};
+}
+
+// FIXME: remove this code duplication with /gapi/test/common/gapi_render_tests.hpp
+void drawMosaicRef(const cv::Mat &mat, const cv::Rect &rect, int cellSz)
+{
+    cv::Rect mat_rect(0, 0, mat.cols, mat.rows);
+    auto intersection = mat_rect & rect;
+
+    cv::Mat msc_roi = mat(intersection);
+
+    bool has_crop_x = false;
+    bool has_crop_y = false;
+
+    int cols = msc_roi.cols;
+    int rows = msc_roi.rows;
+
+    if (msc_roi.cols % cellSz != 0)
+    {
+        has_crop_x = true;
+        cols -= msc_roi.cols % cellSz;
+    }
+
+    if (msc_roi.rows % cellSz != 0)
+    {
+        has_crop_y = true;
+        rows -= msc_roi.rows % cellSz;
+    }
+
+    cv::Mat cell_roi;
+    for (int i = 0; i < rows; i += cellSz)
+    {
+        for (int j = 0; j < cols; j += cellSz)
+        {
+            cell_roi = msc_roi(cv::Rect(j, i, cellSz, cellSz));
+            cell_roi = cv::mean(cell_roi);
+        }
+        if (has_crop_x)
+        {
+            cell_roi = msc_roi(cv::Rect(cols, i, msc_roi.cols - cols, cellSz));
+            cell_roi = cv::mean(cell_roi);
+        }
+    }
+
+    if (has_crop_y)
+    {
+        for (int j = 0; j < cols; j += cellSz)
+        {
+            cell_roi = msc_roi(cv::Rect(j, rows, cellSz, msc_roi.rows - rows));
+            cell_roi = cv::mean(cell_roi);
+        }
+        if (has_crop_x)
+        {
+            cell_roi = msc_roi(cv::Rect(cols, rows, msc_roi.cols - cols, msc_roi.rows - rows));
+            cell_roi = cv::mean(cell_roi);
+        }
+    }
```

Comment:
@TolyaTalamanov I thought there workarounds are not needed for the Mosiac anymore, is it correct?

Reply:
Currently case with fractional mosaic is handled inside opencv backend.
I've opened a ticket to handle it in general case for every backend